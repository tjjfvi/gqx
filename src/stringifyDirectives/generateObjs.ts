
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
