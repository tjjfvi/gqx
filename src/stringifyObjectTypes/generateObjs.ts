
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
