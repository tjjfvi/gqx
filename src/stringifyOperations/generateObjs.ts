
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
