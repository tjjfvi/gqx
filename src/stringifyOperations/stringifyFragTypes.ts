
import { Obj, Context } from ".";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[], ctx: Context) =>
  stringifyNamespace("$$Frag", objs, (o, p) =>
    `type ${p.id.prop} = $$DeepArray<${(p.type in ctx.objectTypes ? `${p.type}$` : "never")}>;`
  )
