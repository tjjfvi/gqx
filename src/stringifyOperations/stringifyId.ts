
import { Id } from ".";

export const stringifyId = (id: Id) =>
  `"${id.type}$${id.prop}"`
