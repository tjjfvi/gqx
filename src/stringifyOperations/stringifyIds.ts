
import { Id } from ".";
import stringifyId from "./stringifyId";

export default (ids: Id[]) =>
  ids.map(id => `class ${stringifyId(id)} { declare private __nominal__: any; }\n`).join("")
