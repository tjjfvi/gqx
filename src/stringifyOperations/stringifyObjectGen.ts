
import { Obj } from "."
import stringifyId from "./stringifyId";
import indent from "../indent";

export default (objs: Obj[]) =>
  `export const $$generateObject = <F extends $$GqxFunc>(f: $$GqxImpl<F>) => ({\n${
    objs.map(o =>
      indent`${o.prop}: {\n${
        o.props.map(p =>
          indent`${p.id.prop}: f(${stringifyId(p.id)}),\n`
        ).join("")
      }},\n`
    ).join("")
  }})`
