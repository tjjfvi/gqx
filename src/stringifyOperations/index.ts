
import { Context } from "..";
import { InputValueDefinitionNode } from "graphql";
import stringifyProps from "./stringifyProps"
import stringifyInputTypes from "./stringifyInputTypes";
import stringifyFragTypes from "./stringifyFragTypes";
import stringifyOutputTypes from "./stringifyOutputTypes";
import stringifyObjectGen from "./stringifyObjectGen";
import generateObjs from "./generateObjs";


interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  wrap: (x: string) => string;
  args: InputValueDefinitionNode[];
}

interface Obj {
  type: string;
  prop: string;
  props: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const [objs] = generateObjs(ctx);

  return [
    stringifyProps(objs),
    stringifyInputTypes(objs),
    stringifyFragTypes(objs, ctx),
    stringifyOutputTypes(objs, ctx),
    stringifyObjectGen(objs),
  ].join("\n\n");
}

