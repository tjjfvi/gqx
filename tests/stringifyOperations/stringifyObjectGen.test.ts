import { generateObjs } from "../../src/stringifyOperations/generateObjs"
import { stringifyObjectGen } from "../../src/stringifyOperations/stringifyObjectGen"

test("", () => {
  let [objs] = generateObjs(utils.getOperations(["stringifyOperations"]))
  let code = stringifyObjectGen(objs)
  expect(code).toParse().toMatchSnapshot()
})
