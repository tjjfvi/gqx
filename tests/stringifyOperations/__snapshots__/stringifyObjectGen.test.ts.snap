// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export const $$generateObject = <F extends $$GqxFunc>(f: $$GqxImpl<F>) => ({
  query: {
    argsObjectArr2D: f(Query$argsObjectArr2D),
    argsObjectArrMM: f(Query$argsObjectArrMM),
    argsObjectArrMN: f(Query$argsObjectArrMN),
    argsObjectArrNM: f(Query$argsObjectArrNM),
    argsObjectArrNN: f(Query$argsObjectArrNN),
    argsObjectMand: f(Query$argsObjectMand),
    argsObjectNull: f(Query$argsObjectNull),
    argsScalarArr2D: f(Query$argsScalarArr2D),
    argsScalarArrMM: f(Query$argsScalarArrMM),
    argsScalarArrMN: f(Query$argsScalarArrMN),
    argsScalarArrNM: f(Query$argsScalarArrNM),
    argsScalarArrNN: f(Query$argsScalarArrNN),
    argsScalarMand: f(Query$argsScalarMand),
    argsScalarNull: f(Query$argsScalarNull),
    noArgsObjectArr2D: f(Query$noArgsObjectArr2D),
    noArgsObjectArrMM: f(Query$noArgsObjectArrMM),
    noArgsObjectArrMN: f(Query$noArgsObjectArrMN),
    noArgsObjectArrNM: f(Query$noArgsObjectArrNM),
    noArgsObjectArrNN: f(Query$noArgsObjectArrNN),
    noArgsObjectMand: f(Query$noArgsObjectMand),
    noArgsObjectNull: f(Query$noArgsObjectNull),
    noArgsScalarArr2D: f(Query$noArgsScalarArr2D),
    noArgsScalarArrMM: f(Query$noArgsScalarArrMM),
    noArgsScalarArrMN: f(Query$noArgsScalarArrMN),
    noArgsScalarArrNM: f(Query$noArgsScalarArrNM),
    noArgsScalarArrNN: f(Query$noArgsScalarArrNN),
    noArgsScalarMand: f(Query$noArgsScalarMand),
    noArgsScalarNull: f(Query$noArgsScalarNull),
  },
  mutation: {
    argsObjectArr2D: f(Mutation$argsObjectArr2D),
    argsObjectArrMM: f(Mutation$argsObjectArrMM),
    argsObjectArrMN: f(Mutation$argsObjectArrMN),
    argsObjectArrNM: f(Mutation$argsObjectArrNM),
    argsObjectArrNN: f(Mutation$argsObjectArrNN),
    argsObjectMand: f(Mutation$argsObjectMand),
    argsObjectNull: f(Mutation$argsObjectNull),
    argsScalarArr2D: f(Mutation$argsScalarArr2D),
    argsScalarArrMM: f(Mutation$argsScalarArrMM),
    argsScalarArrMN: f(Mutation$argsScalarArrMN),
    argsScalarArrNM: f(Mutation$argsScalarArrNM),
    argsScalarArrNN: f(Mutation$argsScalarArrNN),
    argsScalarMand: f(Mutation$argsScalarMand),
    argsScalarNull: f(Mutation$argsScalarNull),
    noArgsObjectArr2D: f(Mutation$noArgsObjectArr2D),
    noArgsObjectArrMM: f(Mutation$noArgsObjectArrMM),
    noArgsObjectArrMN: f(Mutation$noArgsObjectArrMN),
    noArgsObjectArrNM: f(Mutation$noArgsObjectArrNM),
    noArgsObjectArrNN: f(Mutation$noArgsObjectArrNN),
    noArgsObjectMand: f(Mutation$noArgsObjectMand),
    noArgsObjectNull: f(Mutation$noArgsObjectNull),
    noArgsScalarArr2D: f(Mutation$noArgsScalarArr2D),
    noArgsScalarArrMM: f(Mutation$noArgsScalarArrMM),
    noArgsScalarArrMN: f(Mutation$noArgsScalarArrMN),
    noArgsScalarArrNM: f(Mutation$noArgsScalarArrNM),
    noArgsScalarArrNN: f(Mutation$noArgsScalarArrNN),
    noArgsScalarMand: f(Mutation$noArgsScalarMand),
    noArgsScalarNull: f(Mutation$noArgsScalarNull),
  },
})"
`;
